src/core/components/pin.js

```diff
@@ -0,0 +1,438 @@
+'use strict'
+
+const dagPB = require('ipld-dag-pb')
+const DAGNode = dagPB.DAGNode
+const DAGLink = dagPB.DAGLink
+const CID = require('cids')
+const pinSet = require('./pin-set')
+const resolvePaths = require('../utils').resolvePaths
+const promisify = require('promisify-es6')
+const multihashes = require('multihashes')
+const Key = require('interface-datastore').Key
+const each = require('async/each')
+const series = require('async/series')
+const waterfall = require('async/waterfall')
+const parallel = require('async/parallel')
+const until = require('async/until')
+const once = require('once')
+
+function toB58String (hash) {
+  return new CID(hash).toBaseEncodedString()
+}
+
+module.exports = function pin (self) {
+  let directPins = new Set()
+  let recursivePins = new Set()
+  let internalPins = new Set()
```

Comment:
What's the purpose of this Set?  From what I can see it's an in-memory cache of any hashes of pinned DAGNodes that we load or otherwise encounter but I can't see any reference to `internal` as a type in the [pin spec](https://github.com/ipfs/interface-ipfs-core/blob/master/SPEC/PIN.md).

Reply:
@achingbrain Good question. I've only vaguely figured it was to prevent gc as it is not used except to keep a record of some multihashes. I've looked more into it and found:

Internal pins are only updated when flushing/loading pins to/from the datastore. The sources of our pins are:

1) A hash of a totally empty DAGNode. (kindof a zero-hash)
2) Pins are stored as two dags: the dag of recursive pins and dag of direct pins
    - *The hash of those pinset roots are added*
3) When we have > 8192 recursive or direct pins, we shard it into ~256 bins.
    - *The hash of each of those bins are added*
    - If any of those bins have more than 8192 items, we repeat the sharding

This means that when we have < 8192 pins, we have either 2 or 3 multihashes in `internalPins`. When we have more then 8192 pins, we likely have ~ 250. Then it's basically static until we have about 8192 * 256 pins, where it spikes to like 256^2 internal pins.

I think it's intended to keep a reference to prevent some gc during those high-pin count cases but I don't know enough about V8 gc to say. Looking through its uses, I don't think it keeps a ref to anything that isn't referenced somehow else. It might just be a [carry-over from go's impl](https://github.com/ipfs/go-ipfs/blob/master/pin/set.go#L148).

I've tested removing `internalPins` during a 8192 * 256 pin stress test but still hit OOM w/ 2Gb ram. Do you have suggestions what might be going on?

Reply:
@diasdavid The above is the open question I had about V8 garbage collection. I think that `internalPins` was used to prevent collection of pins in the go implementation and was carried over here but I don't know if it does anything in js-land.

Reply:
Friendly ping @diasdavid @achingbrain. I'd like to remove `internalPins` but want another's thought first.